### DISCUSSION TITLE
Повторная инициализация виджета

### DISCUSSION THREAD

**Discussion by AndreyVasilkov** [cdek-it/widget#6] [Q&A] [answered]
Добрый день.
Версия 3.6.0

Создаю объект виджета

```
window.ISDEKWidjetController = new CDEKWidget({
    defaultLocation: "Москва",
    servicePath: "/sdek-widget/",
    tariffs: {
        office: office,
        door: door,
        pickup: pickup
    },
    apiKey: yandexApiKey,
    root: sdekContainerId,
    canChoose:true,
    goods: goods,
    detailAddress: true,
    popup: true,
    onReady: function(){
        console.log("")
    },
    onChoose: function(mode, wat, address) {
        console.log("choose");
    },
});
```
Затем корзина обновляется через ajax (например, смена кол-ва) и инициализация объекта происходит заново, но при это старые слушатели события onReady и onChoose остаются, и соответственно в логе я уже получу несколько выводов choose и ready. Есть ли у объекта CDEKWidget деструктор? в описании к виджету не нашел. Заранее спасибо

**Comment by vermorag** [comment, marked as answer]
Деструктор есть. https://github.com/cdek-it/widget/blob/v3.7.0/dist/cdek-widget.es.d.ts#L129
Однако он функционирует в тестовом режиме. Например - не удаляет карту из DOM (сам яндекс этого не делает). 
Другой вариант - иметь один виджет и обновлять тарифы/офисы/местоположение виджета через
https://github.com/cdek-it/widget/blob/v3.7.0/dist/cdek-widget.es.d.ts#L124
https://github.com/cdek-it/widget/blob/v3.7.0/dist/cdek-widget.es.d.ts#L126
https://github.com/cdek-it/widget/blob/v3.7.0/dist/cdek-widget.es.d.ts#L127
Сброс уже сделанного выбора клиента делается через
https://github.com/cdek-it/widget/blob/v3.7.0/dist/cdek-widget.es.d.ts#L128

Именно таким образом поступает модуль для вордпреса, к примеру.

**Reply by AndreyVasilkov** [reply]
Вариант одного виджета нам не подходит. Возможно ли реализовать в ближайшей версии деструктор, который бы отвязывал слушателей событий onReady и onChoose? Ссылки в ответе сбились

**Reply by vermorag** [reply]
Ссылки исправил. Что имеется в виду под "отвязыванием" слушателей?

**Reply by JuliaPolysalova** [reply]
У меня похожая проблема возникла с использованием метода destroy
после его выполнения, виджет будто продолжает существовать, тк при новой инициализации все действия, прописанные в onReady (вызывала обычный console.log для тестов) выполняются столько раз, сколько был раз инициализирован cdekWidget, хотя перед каждой инициализацией я выполняю destroy

Мы бы с удовольствием использовали 1 виджет, НО как изменить hideDeliveryOptions?
Необходимо, чтобы выбор происходил не внутри виджета, а сверху в контролах (тк в текущей реализации при любом миссклике пвз на карте виджета происходит автоматическое переключение на способ доставки "до двери")
Возможно есть какой-то метод?

И как корректно рассчитать стоимость доставки с надбавкой, если нужно передавать ссылку в servicePath с параметром shipment[originalPrice], а туда полную стоимость, которая может меняться, если изменяется состав товаров/грузов? Тут тоже только переинициализация поможет или есть метод?
